## Setup

Run the code below to load the scrambled message:


In [12]:
import pandas as pd
import re
message = pd.read_csv("https://www.dropbox.com/s/lgpn3vmksk3ssdo/scrambled_message.txt?dl=1")['Word']

In [2]:
message

0                    Koila!
1                     In   
2                     kiew,
3                         a
4                 humble   
               ...         
122                     you
123                 mabugh.
124              ughhh?call
125        meugh.ughhhh!   
126                      K.
Name: Word, Length: 127, dtype: object

In this activity, a "word" refers to any set of characters with no white space, even though they are not truly an English word.  That is, even though many of elements of the scrambled message vector are nonsense, and some have punctuation, you can consider each element to be a "word".

Beware!  The object named `message` is a **pandas Series** of strings.  If you want to use functions that expect a string, rather than a series, remember `.apply()` and `lambda` functions.




## Warm-up exercises

1. How many characters are in the scrambled message?
2544
2. How many of these characters are white space?
1652
3. How many words are in the scrambled message?
127
4. Show all the punctuation marks in the scrambled message.
5. Print out, in all capitals, the longest word in the scrambled message.
KAUDEVILLIANUGH?AOGHAJDBN
6. Print out every piece of a word that starts with the letter "m" and ends with the letter "z" in the scrambled message.
maaz


In [31]:
#how many characters are in the scrambled message
message.str.len().sum()
#how many characters are white space
message.str.count(r'\s').sum()
#how many words
message.str.split().str.len().sum()
#show all punctuation marks
message.str.findall(r'[^\w\s]')
#print out in all captitals the longest word in the scrambled message
message_str = " ".join(message)
words = message_str.split()
longest_word = max(words, key=len)
longest_word.upper()
#print every piece of a word that starts with m and ends with z
re.findall(r'\bm\w*z\b', message_str)

['maaz']



## Decode a message

Complete the following steps to decode the message.  

1. Remove any spaces before or after each word.
2. Any time you see the word "ugh", with any number of h's, followed by a punctuation mark, delete this.
3. No word should be longer than 16 characters. Drop all extra characters beyond 13 off the end of each word.
4. Replace all instances of exactly 2 a's with exactly 2 e's.
5. Replace all z's with t's.
6. Every word that ends in b, change that to a y.  *Hint: look out for punctuation!*
7. Every word that starts with k, change that to a v.  *Hint: look out for capitalization!*
8. Use `.join()` to recombine all your words into a message.
9. Find the movie this quote is from.
V for Vendetta

In [63]:
message_str = " ".join(message)
#remove any spaces before or after each word
no_space = message_str.strip()
#any time the word ugh and punctuation, delete this
no_ugh = re.sub(r'ugh+h*\W', '', message_str)
#no word should be longer than 16 characters
#drop all extra characters beyond 13 off the end of each word
step_3 = " ".join([word[:16] for word in no_ugh.split()]).strip()
#replace exactly 2 a with 2 e
step_4 = re.sub(r'aa', 'ee', step_3)
#replace all z with t
step_5 = re.sub(r'z', 't', step_4)
#every word that ends in b, change that to a y
step_6 = re.sub(r'b\b', 'y', step_5)
#every word that starts with a k, change to a v
lower = step_6.lower()
step_7 = re.sub(r'\bk', 'v', lower)
#use .join() to recombine all words (already done)
print(step_7)

voila! in view, a humble vaudevillianaogh veteran, cast vicariously as both victim and villain by the vicissitudes of fate. this visage, no mere veneer of vanity, is a vestige of the vox populi now vacant, vanished. however, this valorous visitation of a bygone vexation stands vivified, and has vowed to vanquish these venal and virulent vermin, van guarding vice and vouchsafing the violently vicious and voracious violation of volition. the only verdict is vengeance; a vendetta, held as a votive not in vain, for the value and veracity of such shall one day vindicate the vigilant and the virtuous. verily this vichyssoise of verbiage veers most verbose, so let me simply add that its my very good honour to meet you and you may call me v.
voila! in view, a humble vaudevillianaogh veteran, cast vicariously as both victim and villain by the vicissitudes of fate. this visage, no mere veneer of vanity, is a vestige of the vox populi now vacant, vanished. however, this valorous visitation of a b